### PR TITLE
Fix playback of videos in Gallery.

### PR DESCRIPTION
**Pull request by dbrant**
Video playback is currently broken.
This is because our logic for selecting the best possible derivative for playback was not smart enough, and was blindly selecting the "last" derivative.
The list of derivatives can change, and indeed it has changed to include an unplayable derivative as the "last" one.
This improves the logic of selecting the proper derivative.

https://phabricator.wikimedia.org/T352315

### CODE REVIEW

File: app/src/main/java/org/wikipedia/gallery/ImageInfo.kt [hunk 1]
```diff
@@ -46,16 +46,25 @@ class ImageInfo {
     val width = 0
     val height = 0
 
-    val bestDerivative get() = derivatives.lastOrNull()
+    fun getBestDerivativeForSize(widthDp: Int): Derivative? {
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/gallery/ImageInfo.kt:49]: Can we add `@Suppress("KotlinConstantConditions")` for this function?

File: app/src/main/java/org/wikipedia/gallery/ImageInfo.kt [hunk 2]
```diff
@@ -46,16 +46,25 @@ class ImageInfo {
...
+        var derivative: Derivative? = null
+        derivatives.forEach {
+            if (it.width > 0 && it.width < widthDp) {
+                if (derivative == null || it.width > derivative!!.width) {
+                    derivative = it
+                }
+            }
+        }
```
Comment (cooltey) [inline app/src/main/java/org/wikipedia/gallery/ImageInfo.kt:57]: Would it be better to use `filter` and `mayByOrNull`?
```
derivatives.filter { it.width < widthDp && it.width < widthDp }.maxByOrNull { it.width }
```

I tried a couple of times and it looks like using `filter` and `maxByOrNull` can mostly improve the efficiency
```
forEach                581.093us
filter+mayByOrNull     53.75us

forEach                62.032us
filter+mayByOrNull     67.604us

forEach                51.51us
filter+mayByOrNull     50.781us

forEach                118.489us
filter+mayByOrNull     57.239us
```

**Review by cooltey — CHANGES_REQUESTED** [review]

File: app/src/main/java/org/wikipedia/gallery/ImageInfo.kt (hunk 1 above)
Comment (dbrant) [reply app/src/main/java/org/wikipedia/gallery/ImageInfo.kt:50]: of course!

File: app/src/main/java/org/wikipedia/gallery/ImageInfo.kt (hunk 2 above)
Comment (cooltey) [reply app/src/main/java/org/wikipedia/gallery/ImageInfo.kt]: Thanks for the information, that explains the benefit of using the original logic you put in the code! I checked the library and you are correct!

**Pull request merged**